health checker/ auto-retry bg sync + adjusted timeouts

creviera opened the pull request:
# Description

Resolves https://github.com/freedomofpress/securedrop-client/issues/610
Resolves #181

* we no longer ask the user to manually refresh if a bg sync job fails, instead we auto retry in the background (this is what resolves #610)
* adds a `manual_refresh` option to `sync_api` so that if a user clicks the refresh icon and the sync fails, we show an error in the status bar with a retry link
* bumps the default timeout to 60 seconds used for login and delete (this is what resolves #181)
* queue jobs still use a default of 5 seconds for timeouts, so if the server cannot be reached, we will show the error immediately to the user if they click on the refresh icon

 - [x] I have tested these changes in the appropriate Qubes environment

eloquence commented:
We discussed this at sprint planning today; this does not mitigate the observed problem, so Allie may close the PR, or amend if it is worth increasing the timeout regardless.

creviera commented:
"The SecureDrop server cannot be reached" with Retry link continues to intermittently appear even with this extended timeout. This tells us that the issue is not a `RequestTimeoutError`, which means it has to be an `ApiInaccessibleError` (see https://github.com/freedomofpress/securedrop-client/blob/ff394d511497ff462294c5434c1ccf857e1a235f/securedrop_client/queue.py#L120-L123)

Also, note that if the API were None then we wouldn't see that Retry link, so that is not the issue: https://github.com/freedomofpress/securedrop-client/blob/master/securedrop_client/logic.py#L275

We already automatically try 5 times in a row to resend a request from the queue if we see `ApiInaccessibleError` but the retries are very quickly happening back-to-back. One mitigation might be to space out our retries or wait X seconds and auto retry again before pausing the queue and asking for user intervention to fix the network issue. This is a bit like a health checker (see https://github.com/freedomofpress/securedrop-client/issues/491) but it would be less advanced. We could start off by just auto-retrying once after say 30 seconds, specifically for `ApiInaccessableErrors` since we're seeing this frequently.

redshiftzero commented:
are you sure it's an `ApiInaccessibleError`? can you confirm via the debug logging?

creviera commented:
> are you sure it's an ApiInaccessibleError? can you confirm via the debug logging?

Ah actually I just realized since `MetadataSyncJob` isn't implemented yet that the errors would have to be coming from `MessageDownloadJob` or `ReplyDownloadJob` which are triggered during a sync. So either of those jobs could be seeing timeout errors. 

If we increased the timeouts for those jobs that would actually work as a quick mitigation if it is a timeout issue. If we're seeing frequent `ApiInaccessibleError`s then I think waiting and auto retrying again makes sense.

I can verify what the issue is once I get to my Qubes station.

creviera commented:
It does seem like increasing the `default_request_timeout` [here]( https://github.com/freedomofpress/securedrop-client/pull/612/files) (which should be applied to both `MessageDownloadJob` and `ReplyDownloadJob`) to a very large value doesn't help this issue, which is why I'm thinking the api is just temporarily inaccessible. I will confirm shortly.

creviera commented:
Resolves https://github.com/freedomofpress/securedrop-client/issues/181

creviera commented:
I updated the PR description to reflect recent changes and added tests after rebasing to include the metadata sync job changes, so this is ready for review!

redshiftzero commented on securedrop_client/queue.py:
```diff
@@ -160,8 +159,13 @@ def logout(self) -> None:
+        # General requests includes:
+        #       FileDownloadJob
+        #       MessageDownloadJob
+        #       ReplyDownloadJo
```
for `FileDownloadJob, MessageDownloadJob, ReplyDownloadJob` we're using `DownloadJob._get_realistic_timeout` so we might want to remove them from this list to prevent confusion

creviera replied:
good point, removed comment

redshiftzero commented on securedrop_client/logic.py:
```diff
@@ -304,11 +304,16 @@ def completed_api_call(self, thread_id, user_callback):
         storage.mark_all_pending_drafts_as_failed(self.session)
-        self.api = sdclientapi.API(self.hostname, username, password, totp, self.proxy)
+        self.api = sdclientapi.API(
+            self.hostname, username, password, totp, self.proxy, default_request_timeout=60)
```
Right now this is altering the default request timeouts for deletions but only temporarily until we implement the `DeletionJob` ✅

creviera replied:
Yep, this alters the timeout used during login (`self.api.authenticate`) deletion (`self.api.delete_source`), and logout (`self.api.logout`)

redshiftzero commented on securedrop_client/logic.py:
```diff
@@ -412,8 +421,17 @@ def on_sync_success(self) -> None:
+        a background task.
+        """
+        logger.debug('The SecureDrop server cannot be reached due to Error: {}'.format(result))
+        self.sync_api()
```
I can see this makes sense if the reason for the sync errors is that `ApiInaccessibleError` is being raised - when you investigated, _was_ it `ApiInaccessibleError` that was the cause? If it is, I'm not following why the `ApiInaccessibleError` is raised. My expectation is that it should only occur when one of the following is true:

* User's token has expired
* User's token is somewhere otherwise invalid
* There is a bug

creviera replied:
When we run a sync, we kick off a `MetadataSyncJob` which on success will kick off `MessageDownloadJob`s, `ReplyDownloadJob`s, or `FileDownloadJob`s. On failure (`ApiInaccessibleError` or `RequestTimeoutError` or any kind of failure) we used to _always_ show "The SecureDrop server cannot be reached" error with the Retry link. Now, on failure, we check if the sync ran because of a user action (clicking refresh) or not. If the action was not from a user action, we don't ask the user to automatically retry.

Looking at this more, I'm realizing that I forgot to add the code that checks if the exception in that comes back via the failure signal of `MetadataSyncJob` is because we could not reach the server. My change introduces retrying for any kind of error which is problematic.

Overall, I'm curious if you think it's a good approach to periodically retry connecting to the server when it cannot be reached until we can successfully metadata sync. Let's discuss this more because I may be going in a direction you disagree with.

creviera replied:
> when you investigated, was it ApiInaccessibleError that was the cause?

When I investigated I saw that for any kind of error during sync_api we would show "The SecureDrop server cannot be reached" error with the Retry link. So the answer is sometimes.

One thing we still need to add is increasing the timeouts for all jobs before we pause the queue in case 5 seconds is not long enough to download messages, metadata for file downloads, etc.

redshiftzero replied:
so in latest `master`, calling `sync_api` is now just going to add a new `MetadataSyncJob` to the queue, it won't restart the queue, and since the queue will be paused if a `ApiInaccessibleError` or `RequestTimeoutError` occurred, that job will get automatically retried upon queue restart, so I don't _think_ we need this logic. Let me know if you disagree. If the sync job fails for some other reason, it means something else is wrong (like https://github.com/freedomofpress/securedrop-proxy/issues/13).

> I'm curious if you think it's a good approach to periodically retry connecting to the server when it cannot be reached until we can successfully metadata sync.

I do think we want this but I don't think we want to do this as part of this ticket (I think that's #491) and I do also agree with the direction of eventually using the MetadataSyncJob for #491.

creviera replied:
you are right, resuming queue is what was intended, just fixed this.

redshiftzero commented on securedrop_client/logic.py:
```diff
@@ -412,8 +421,17 @@ def on_sync_success(self) -> None:
+        a background task.
+        """
+        logger.debug('The SecureDrop server cannot be reached due to Error: {}'.format(result))
+        self.sync_api()
```
recapping from synchronous chat just now: we are going to remove the self.sync_api() call here and instead resume the queues. By doing so, we are also closing #491, and we are going to treat the `MetadataSyncJob` as our network health checker. The queue pause / retry error message will continue to appear as per the acceptance criteria in that ticket until the network action is confirmed to succeed. The removal of that error message will be doing in the `on_sync_success` callback.

redshiftzero commented on securedrop_client/logic.py:
```diff
@@ -374,9 +379,13 @@ def sync_api(self):
+            if manual_refresh:
+                job.failure_signal.connect(self.on_refresh_failure, type=Qt.QueuedConnection)
+            else:
+                job.failure_signal.connect(self.on_sync_failure, type=Qt.QueuedConnection)
```
Here we can (tbh I think this is optional now but I'll let you decide if you want to add it) increase the number of tries for the sync job _not_ triggered by the user since in that case it doesn't really matter if they get rapid feedback. We can do this by setting `remaining_attempts` explicitly on the job.

creviera commented:
Resolves https://github.com/freedomofpress/securedrop-client/issues/491

creviera commented:
I just want to mention that when you're ready to re-review that the default api timeouts are:

* reply_source: 5 seconds (unchanged)
* remove_star: 5 seconds (unchanged)
* add_star: 5 seconds (unchanged)
* download jobs (except for MetadataSyncJob): adjusted realistic timeout based on download size (unchanged)
* MetadataSyncJob get_sources: 5 seconds -> changed to 20 seconds
* MetadataSyncJob get_submissions: 5 seconds -> changed to 20 seconds
* MetadataSyncJob get_all_replies: 5 seconds -> changed to 20 seconds

Changing the MetadataSyncJob api requests from 5 to 20 seconds makes it so I basically never see a network error. I tested this by running the client for 3 hours. At one point I saw the network error message and did nothing and later saw that it had resolved itself. This demonstrated that if a user walks away or is busy reading/ editing a file in a different vm or something, and there are network errors, that they can potentially resolve themselves. So when a user goes back to the client they don't have to see the a red error message that requires a manual retry. 

Changing the MetadataSyncJob api requests from 5 to 20 seconds does mean that a refresh will take up to 1 minute before telling a user that the server could not be reached. However, other operations will tell the user that the server could not be reached within 25 seconds. So that is the tradeoff, and I opened a followup issue to adjust all the defaults in a smarter way, similar to what we did for download jobs: https://github.com/freedomofpress/securedrop-client/issues/648

creviera commented:
Latest commit removes extra exception logging now that metadata sync is a queue job in order to avoid repetitive retry messages flooding the logs.

redshiftzero commented on securedrop_client/logic.py:
```diff
@@ -401,6 +417,8 @@ def on_sync_success(self) -> None:
             * Download new messages and replies
             * Update missing files so that they can be re-downloaded
         """
+        self.set_status('')  # remove any permanent error status message
```
should be `self.gui.clear_error_status`? it looks like `self.set_status` is modifying `TopPane.activity_status_bar` and `self.gui.clear_error_status` is modifying `TopPane.error_status_bar`

creviera replied:
fixed!

redshiftzero replied:
ah so this should be `clear_error_status`, `update_error_status` will display the error status bar but with no message

redshiftzero commented on securedrop_client/logic.py:
```diff
@@ -439,6 +470,7 @@ def on_update_star_success(self, result) -> None:
         """
         After we star a source, we should sync the API such that the local database is updated.
         """
+        self.set_status('')  # remove any permanent error status message
```
I see why you're doing this: the next job that _might_ succeed isn't necessarily a `MetadataSyncJob`, hence needing `self.set_status` sprinkled in our success callbacks. We should put this somewhere generic, some options:

1. we reach into the gui object from the queue - technically "works" but this transitive navigation is poor design
2. we emit a separate signal indicating that the error status should be cleared - I guess that works too but I don't love it - we already have a success signal getting emitted and it would unfortunate to complicate the design of the queue for this purpose
3. we create a decorator `@success_callback` and put any per-success logic in there along with our explanation of why it's needed, and then decorate any success callbacks with it. It's the same size diff, but it's a bit more extensible. We could also do the same for the failure callback to ensure that all failures have a log line (since most of our failure callbacks have a debug line to log a failure, we could put that in a generic place too to ensure that we're logging these failures). 

anyway typing this out I'm liking 3 but also feeling like this should be a followup - what do you think @creviera ?

creviera replied:
Yes, I agree with #3 and it being a followup issue

creviera replied:
I agree with #3 and captured this in a followup issue: https://github.com/freedomofpress/securedrop-client/issues/651

redshiftzero commented on securedrop_client/logic.py:
```diff
@@ -401,6 +417,8 @@ def on_sync_success(self) -> None:
             * Download new messages and replies
             * Update missing files so that they can be re-downloaded
         """
+        self.gui.clear_error_status()  # remove any permanent error status message
```
this call is also needed on: 

on_file_download_success
